### PR TITLE
fixed the issue of DPO trainer that using one node and mutiple GPUs and set the device_map='auto'

### PR DESCRIPTION
**Pull request by pengwei715**
# What does this PR do?

Fixes 
https://github.com/huggingface/trl/issues/958
https://github.com/huggingface/trl/issues/1399

It's a very strong assumption that `tr_loss` and `tr_loss_step` are on the same device. `arg.device` may not be as same as the current device. 

For example, the `dpo_trainer` is child class of `trainer`. If you are running a DPO training job with one node multiple GPUs. If you set the `device_map='auto'`. The `tr_loss_step` could be on different device after computing the `dpo_loss`.  

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@lvwerra 
@kashif 
@younesbelkada 
@guy1992l

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

Let's just get a double check from @younesbelkada as it involves TRL, in case there's other PRs / issues open related to this

**Review by younesbelkada — COMMENTED** [review]
Thanks for the fix @pengwei715 ! 
IMO the fix should live in TRL, can you try to set the step loss on the correct device here: https://github.com/huggingface/trl/blob/6cfa5cfc817be0f8acf30f8158a4a0ea244dc988/trl/trainer/dpo_trainer.py#L1084 and see if it works? If that's the case would you like to open the fix on TRL directly? 🙏 Thanks!

**Comment by guy1992l** [comment]
Hi @younesbelkada 

Thanks for your reply! I completely understand the logic of "it's related to `trl` then the fix should be in `trl`", but just to be clear: It should be the case if and only if there is no chance that `tr_loss` and `tr_loss_step` can be on separate devices as it is the responsibility of the one who calculate the loss to put it on the right device. If so, maybe we should add an assert:

```python
assert tr_loss.device == tr_loss_step.device, f"Calculated loss must be on the original device: {tr_loss.device} but device in use is {tr_loss_step.device}"
``` 

Regarding the fix, maybe just to spare our time you know how to get the original `tr_loss.device` back in `DPOTrainer` where you point out the fix should be? I saw it is taken from `args.device`, is `args` available there?

**Comment by younesbelkada** [comment]
Hi @guy1992l !
Yes having an assert could work ! You can retrieve `args` from `self` directly: https://github.com/huggingface/transformers/blob/c852d4fba6a46f8b71c608041a10691b66f174e5/src/transformers/trainer.py#L363 let me know if that works !

**Comment by pengwei715** [comment]
Hi @younesbelkada, @amyeroberts, @guy1992l 

Thank you very much for the feedback. I will test it out. And Let you know.

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1983,7 +1983,9 @@ def _inner_training_loop(
                     # if loss is nan or inf simply add the average of previous logged losses
                     tr_loss += tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)
                 else:
-                    assert tr_loss.device == tr_loss_step.device, f"Calculated loss must be on the original device: {tr_loss.device} but device in use is {tr_loss_step.device}"
+                    assert (
+                        tr_loss.device == tr_loss_step.device
+                    ), f"Calculated loss must be on the original device: {tr_loss.device} but device in use is {tr_loss_step.device}"
```
Comment (guy1992l) [inline src/transformers/trainer.py:1988]: please verify the `x.device` is printable so it will have a message that makes sense and not a memory address of the object :)

**Review by guy1992l — COMMENTED** [review]
a tiny concern and also, please update the description to mach Huggingface's requirements now that the PR has shifted towards a "side quest" of this fix.

File: src/transformers/trainer.py (hunk 1 above)
Comment (pengwei715) [reply src/transformers/trainer.py:1988]: Confirmed. tested. it's  printable.

**Comment by pengwei715** [comment]
Hi @younesbelkada , @amyeroberts , @guy1992l ,

This PR is doing the assertion checking to make sure the `tr_loss` and `tr_loss_step` are in the same device before updating. 

For the `dpo_trainer`, I have opened another PR here to address the issue inside of the `DPOTrainer` class. The PR is here. https://github.com/huggingface/trl/pull/1439. 

Both are tested. 
Please let me know your thoughts. Thanks.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1983,6 +1983,9 @@ def _inner_training_loop(
                     # if loss is nan or inf simply add the average of previous logged losses
                     tr_loss += tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)
                 else:
+                    assert (
```
Comment (younesbelkada) [inline src/transformers/trainer.py:1986]: We don't have any `assert` in our codebase, we test a condition and we raise a specific error if that condition is not met

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1983,6 +1983,9 @@ def _inner_training_loop(
...
+                        tr_loss.device == tr_loss_step.device
+                    ), f"Calculated loss must be on the original device: {tr_loss.device} but device in use is {tr_loss_step.device}"
```
Comment (younesbelkada) [inline src/transformers/trainer.py:1988]:
```suggestion
                    if tr_loss.device != tr_loss_step.device:
                        ValueError(f"Calculated loss must be on the original device: {tr_loss.device} but device in use is {tr_loss_step.device}")
```

**Review by younesbelkada — COMMENTED** [review]
Thanks so much @pengwei715 I left one suggestion !

**Comment by younesbelkada** [comment]
For the next reveiwers - the core issue should be fixed with https://github.com/huggingface/trl/pull/1439 - this PR simply checks if the losses are on the correct device, this should be safe IMO as only sub-classed trainers would potentially face this issue and the training would fail anyway if the devices are not the same

File: src/transformers/trainer.py (hunk 2 above)
Comment (pengwei715) [reply src/transformers/trainer.py]: Hi @younesbelkada,  Thank you very much. I will keep that in mind in the future PR.

**Review by muellerzr — APPROVED** [review]
Thanks! I agree that the fix should be over in `trl`, but it makes sense that we ensure that the devices are the same in this case otherwise raise a better err for users. Thanks!

(You may need to run `make style; make quality`

**Comment by pengwei715** [comment]
> Thanks! I agree that the fix should be over in `trl`, but it makes sense that we ensure that the devices are the same in this case otherwise raise a better err for users. Thanks!
> 
> (You may need to run `make style; make quality`

@muellerzr , Thank you very much. Let me do that.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1983,6 +1983,10 @@ def _inner_training_loop(
                     # if loss is nan or inf simply add the average of previous logged losses
                     tr_loss += tr_loss / (1 + self.state.global_step - self._globalstep_last_logged)
                 else:
+                    if tr_loss.device != tr_loss_step.device:
+                        ValueError(
```
Comment (amyeroberts) [inline src/transformers/trainer.py:1987]: I think we have to raise here - otherwise I don't think anything will happen 

```suggestion
                        raise ValueError(
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by pengwei715** [comment]
> Reference in new

Silly me. Thanks. @amyeroberts

**Comment by pengwei715** [comment]
Hi @amyeroberts , It seems the `tests_hub` failed because of too many requests. Could you help to rerun the failed tests. Thank you very much.

**Comment by pengwei715** [comment]
Hi @amyeroberts , It seems this PR is ready to merge. Please let me know if more changes are still needed. Thanks.

**Review by amyeroberts — APPROVED** [review]
Thanks!
